crates/lib3h/src/rrdht_util/r_value.rs

```diff
@@ -0,0 +1,335 @@
+
+/// builder for our knowledge of a single agent's storage / reachability
+pub struct RValuePeerRecord {
+    storage_arc: Arc,
+    uptime_0_to_1: f64,
+}
+
+impl Default for RValuePeerRecord {
+    fn default() -> Self {
+        Self {
+            storage_arc: Arc::new(0.into(), 0),
+            uptime_0_to_1: 0.0,
+        }
+    }
+}
+
+impl RValuePeerRecord {
+    /// what arc is this agent claiming to store
+    pub fn storage_arc(mut self, storage_arc: Arc) -> Self {
+        self.storage_arc = storage_arc;
+        self
+    }
+
+    /// what is this agent's uptime
+    /// - 0 = never reachable
+    /// - 1 = always reachable
+    pub fn uptime_0_to_1(mut self, uptime_0_to_1: f64) -> Self {
+        self.uptime_0_to_1 = uptime_0_to_1;
+        self
+    }
+}
+
+/// builder for a set of peers we know about
+/// this should generally either be our entire storage arc of peer info
+/// or, if it is huge, small sample arc
+pub struct RValuePeerRecordSet {
+    arc_of_included_peer_records: Arc,
+    peer_records: Vec<RValuePeerRecord>,
+}
+
+impl Default for RValuePeerRecordSet {
+    fn default() -> Self {
+        Self {
+            arc_of_included_peer_records: Arc::new(0.into(), 0),
+            peer_records: Vec::new(),
+        }
+    }
+}
+
+impl RValuePeerRecordSet {
+    /// the arc we are representing with this PeerRecordSet
+    pub fn arc_of_included_peer_records(mut self, arc_of_included_peer_records: Arc) -> Self {
+        self.arc_of_included_peer_records = arc_of_included_peer_records;
+        self
+    }
+
+    /// push a new PeerRecord into this record set
+    pub fn push_peer_record(mut self, peer_record: RValuePeerRecord) -> Self {
+        if !self
+            .arc_of_included_peer_records
+            .contains_location(peer_record.storage_arc.center())
+        {
+            panic!("peer record does not fit within arc_of_included_peer_records");
+        }
+        self.peer_records.push(peer_record);
+        self
+    }
+}
+
+/// Interpolate qualified redundancy in the network given our current
+/// knowledge of peers' claimed storage arcs.
+/// For a given storage arc (should match *this* agent's storage arc),
+/// if all the peers we know about are actually storing what they claim to be
+/// we can interpolate what the whole network "R" value is
+/// also taking into account our experience of those agents' reachability.
+pub fn interpolate_r_value_for_given_arc(peer_record_set: &RValuePeerRecordSet) -> f64 {
+    let mut running_avg_coverage = 0.0_f64;
+
+    for (i, record) in peer_record_set.peer_records.iter().enumerate() {
+        // what is the total coverage factor for this record?
+        let coverage =
+            (record.storage_arc.length() as f64 / ARC_LENGTH_MAX as f64) * record.uptime_0_to_1;
+
+        // mix it into our running average
+        running_avg_coverage = (running_avg_coverage * i as f64 + coverage) / (i as f64 + 1.0);
```

Comment:
I have a suggestion for a different approach to this calculation (based on previous experience trying this in a similar context)...
Welford Stat Tuples (WST):
  - Provides for accumulation of count, expected value, and a variance measure. And it may be really (um, eventually REALLY) useful to capture the variance in addition to the expected coverage.
  - WSTs allow for almost-lossless O(1) updates with additions/deletions. The expected value method currently used will experience significant loss of precision because it will accumulate floating point errors. (you'd think they'd average out, but they don't... and the * i and / i just make it worse unfortunately)
  - WSTs are associative & commutative. Thus (in addition to single additions/deletions) partial statistical results can be aggregated (or dis-aggregated) without loss. This means, we could optimize this calculation when/where DHT shards overlap (e.g. based on a binary-based hash location hierarchies) by sharing partial results... This is a future optimization I want to discuss with you @neonphog, but WSTs will prepare us well.

Reply:
This does not resolve easily on google ;p - I'm not against it, but it doesn't seem like an MVP thing. This code is just to get us going, we should feel free to update our algorithm whenever there is need.

Reply:
Agreed.

And I totally should've linked something useful (but was on a plane without wifi when responding)!

https://www.johndcook.com/blog/standard_deviation/